pkg/metadata/manager.go

```diff
@@ -0,0 +1,374 @@
+	shutdown chan struct{}
+	// Interface to do file operations
+	fschecker fs
+	// Interface to do cgroups related operations
+	cgroupsChecker cgroup
+}
+
+type PodMetadata struct {
+	Name      string
+	Namespace string
+	IPs       []string
+}
+
+const (
+	PodAddEvent = iota
+	PodUpdateEvent
+	PodDeleteEvent
+	PodMetadataEvent
+)
+
+func NewCgroupManager() *CgroupManager {
+	m := &CgroupManager{
+		podMetadataById:           make(map[string]*podMetadata),
+		containerMetadataByCgrpId: make(map[uint64]*containerMetadata),
+		podEvents:                 make(chan podEvent, podEventsChannelSize),
+		shutdown:                  make(chan struct{}),
+	}
+	m.fschecker = fsImpl{}
+	m.cgroupsChecker = cgroupImpl{}
+	m.checkCgroupPath = new(sync.Once)
+
+	m.enable()
+
+	return m
+}
+
+func (m *CgroupManager) OnAddPod(pod *v1.Pod) {
+	if pod.Spec.NodeName != nodetypes.GetName() {
+		return
+	}
+	m.podEvents <- podEvent{
+		pod:       pod,
+		eventType: PodAddEvent,
+	}
+}
+
+func (m *CgroupManager) OnUpdatePod(oldPod, newPod *v1.Pod) {
+	if newPod.Spec.NodeName != nodetypes.GetName() {
+		return
+	}
+	m.podEvents <- podEvent{
+		pod:       newPod,
+		oldPod:    oldPod,
+		eventType: PodUpdateEvent,
+	}
+}
+
+func (m *CgroupManager) OnDeletePod(pod *v1.Pod) {
+	if pod.Spec.NodeName != nodetypes.GetName() {
+		return
+	}
+	m.podEvents <- podEvent{
+		pod:       pod,
+		eventType: PodDeleteEvent,
+	}
+}
+
+// GetParentPodMetadata returns parent pod metadata for the given container
+// cgroup id in case of success, or nil otherwise.
+func (m *CgroupManager) GetParentPodMetadata(cgroupId uint64) *PodMetadata {
+	if !m.enabled {
+		return nil
+	}
+	podMetaOut := make(chan *PodMetadata)
+
+	m.podEvents <- podEvent{
+		cgroupId:       cgroupId,
+		eventType:      PodMetadataEvent,
+		podMetadataOut: podMetaOut,
+	}
+	select {
+	// We either receive pod metadata, or zero value when the channel is closed.
+	case pm := <-podMetaOut:
+		return pm
+	}
+}
+
+// Close should only be called once from daemon close.
+func (m *CgroupManager) Close() {
+	close(m.shutdown)
+}
+
+type podUID = string
+
+type podMetadata struct {
+	name       string
+	namespace  string
+	id         string
+	ips        []string
+	containers map[string]struct{}
+}
+
+type containerMetadata struct {
+	cgroupId    uint64
+	cgroupPath  string
+	parentPodId string
+}
+
+type podEvent struct {
+	pod            *v1.Pod
+	oldPod         *v1.Pod
+	cgroupId       uint64
+	eventType      int
+	podMetadataOut chan *PodMetadata
+}
+
+type fs interface {
+	Stat(name string) (os.FileInfo, error)
+}
+
+type cgroup interface {
+	GetCgroupID(cgroupPath string) (uint64, error)
+}
+
+type fsImpl struct{}
+
+func (f fsImpl) Stat(name string) (os.FileInfo, error) {
+	return os.Stat(name)
+}
+
+type cgroupImpl struct{}
+
+func (c cgroupImpl) GetCgroupID(cgroupPath string) (uint64, error) {
+	return cgroups.GetCgroupID(cgroupPath)
+}
+
+func (m *CgroupManager) enable() {
+	go m.processPodEvents()
```

Comment:
I've made a small change locally to start the goroutine only after the `m.enabled` flag is set, and move the common code to `initManager(...)`. So both manager and manager_test call into the `initManager(...)`. 

<details><summary>diff</summary>
<p>

diff --git a/pkg/metadata/manager.go b/pkg/metadata/manager.go
index 38a9c2d8e1..c122c2c7df 100644
--- a/pkg/metadata/manager.go
+++ b/pkg/metadata/manager.go
@@ -73,19 +73,7 @@ const (
 )

 func NewCgroupManager() *CgroupManager {
-       m := &CgroupManager{
-               podMetadataById:           make(map[string]*podMetadata),
-               containerMetadataByCgrpId: make(map[uint64]*containerMetadata),
-               podEvents:                 make(chan podEvent, podEventsChannelSize),
-               shutdown:                  make(chan struct{}),
-       }
-       m.fschecker = fsImpl{}
-       m.cgroupsChecker = cgroupImpl{}
-       m.checkCgroupPath = new(sync.Once)
-
-       m.enable()
-
-       return m
+       return initManager(fsImpl{}, cgroupImpl{}, podEventsChannelSize)
 }

 func (m *CgroupManager) OnAddPod(pod *v1.Pod) {
@@ -188,8 +176,24 @@ func (c cgroupImpl) GetCgroupID(cgroupPath string) (uint64, error) {
        return cgroups.GetCgroupID(cgroupPath)
 }

+func initManager(fs fs, cg cgroup, channelSize int) *CgroupManager {
+       m := &CgroupManager{
+               podMetadataById:           make(map[string]*podMetadata),
+               containerMetadataByCgrpId: make(map[uint64]*containerMetadata),
+               podEvents:                 make(chan podEvent, channelSize),
+               shutdown:                  make(chan struct{}),
+       }
+       m.fschecker = fs
+       m.cgroupsChecker = cg
+       m.checkCgroupPath = new(sync.Once)
+
+       m.enable()
        go m.processPodEvents()
+
+       return m
+}
+
</p>
</details>